dandischema/tests/test_models.py

```diff
@@ -689,3 +692,53 @@ def test_embargoedaccess() -> None:
             )
         ]
     )
+
+
+def _non_contact_person_roles_args() -> List[List[RoleType]]:
+    """
+    Return a list of role lists that each do not contain the role of a contact person
+    """
+    return [[], [RoleType.Author, RoleType.DataCurator], [RoleType.Funder]]
+
+
+def _contact_person_roles_args() -> List[List[RoleType]]:
+    """
+    Return a list of role lists that each contains the role of a contact person
+    """
+    return [
+        role_lst + [RoleType.ContactPerson]
+        for role_lst in _non_contact_person_roles_args()
+    ]
+
+
+class TestContributor:
+    @pytest.mark.parametrize("roles", _contact_person_roles_args())
+    def test_contact_person_with_email(self, roles: List[RoleType]) -> None:
+        """
+        Test creating a `Contributor` instance as a contact person with an email
+        """
+        Contributor(email="nemo@dandiarchive.org", roleName=roles)
+
+    @pytest.mark.parametrize("roles", _contact_person_roles_args())
+    def test_contact_person_without_email(self, roles: List[RoleType]) -> None:
+        """
+        Test creating a `Contributor` instance as a contact person without an email
+        """
+        with pytest.raises(
+            pydantic.ValidationError, match="Contact person must have an email address"
+        ):
+            Contributor(roleName=roles)
+
+    @pytest.mark.parametrize("roles", _non_contact_person_roles_args())
+    def test_non_contact_person_with_email(self, roles: List[RoleType]) -> None:
+        """
+        Test creating a `Contributor` instance as a non-contact person with an email
+        """
+        Contributor(email="nemo@dandiarchive.org", roleName=roles)
```

Comment:
IMHO there is little point to separate out this one from `test_contact_person_with_email` -- we do not need to do full combinatoric combination 2x2:  let's merge into the 

```python

    @pytest.mark.parametrize("roles", _contact_person_roles_args() + _non_contact_person_roles_args())
    def test_person_with_email(self, roles: List[RoleType]) -> None:
        """
        Test creating a `Contributor` instance with an email
        """
        Contributor(email="nemo@dandiarchive.org", roleName=roles)
```

NB subject to also getting rid unnecessary function invocation - could be just const lists of roles.

Reply:
I separated the tests cases to make it easier to locate bugs, but this way is indeed more compact.